### PR TITLE
Check resource limits for free plan before resources upload

### PR DESCRIPTION
**Pull request by tonynovi**

### CODE REVIEW

File: src/tree_panel.py [hunk 1]
```diff
@@ -26,8 +26,10 @@
 import sys
 import json
 import functools
+import requests
```
Comment (5Nap) [inline src/tree_panel.py:29]: requests is not a standard module

File: src/tree_panel.py [hunk 2]
```diff
@@ -1064,11 +1066,103 @@ def create_group(self):
...
+            if not SUCCESS_KEY in quota_response.json():
+                qgisLog('Unable to parse quota response.', level=CompatQgisMsgLogLevel.Info)
+                return False
+            qgisLog(quota_response.json()[MESSAGE_KEY],
+                    level=CompatQgisMsgLogLevel.Info if quota_response.json()[SUCCESS_KEY] else CompatQgisMsgLogLevel.Critical)
+            return quota_response.json()[SUCCESS_KEY]
+        return True
+
+    def get_items_to_upload_count(self, layers, is_project=False):
+        items_count = {
```
Comment (5Nap) [inline src/tree_panel.py:1132]:
```
import collections
items_count = collections.Counter
```

File: src/tree_panel.py [hunk 3]
```diff
@@ -1064,11 +1066,103 @@ def create_group(self):
...
+        )
+        if result == QMessageBox.Yes:
+            webbrowser.open('https://my.nextgis.com/subscription/update')
     def import_qgis_project(self):
         sel_index = self.trvResources.selectionModel().currentIndex()
 
-        current_project = QgsProject.instance()
+        current_project = CompatQgis.layers_registry()
         current_project_title = current_project.title()
+        layers_to_check = [layer for name, layer in current_project.mapLayers().items()]
```
Comment (5Nap) [inline src/tree_panel.py:1161]: `layer for layer in mapLayers().values()`

File: src/tree_panel.py [hunk 4]
```diff
@@ -1114,6 +1208,11 @@ def import_layers(self):
             if len(qgs_map_layers) == 0: # just in case if checkImportActionsAvailability() works incorrectly
                 return
 
+        items_to_upload_count = self.get_items_to_upload_count(qgs_map_layers, False)
```
Comment (5Nap) [inline src/tree_panel.py:1211]: Combining it into a single method would make it cleaner (you did the same two-steps action at least twice)
```
if not self.subscription_limits_fit_layers(layers):
    self.display_update_subscription....
```

File: src/tree_panel.py [hunk 5]
```diff
@@ -1064,11 +1066,103 @@ def create_group(self):
...
+            'vector_layer': 0,
+            'webmap': 0
+        }
+        for layer in layers:
+            if isinstance(layer, QgsRasterLayer):
+                items_count['raster_layer'] += 1
+            elif isinstance(layer, QgsVectorLayer):
+                items_count['vector_layer'] += 1
+        if is_project:
+            items_count['webmap'] = 1
```
Comment (5Nap) [inline src/tree_panel.py:1143]: Mixing of different entities

File: src/tree_panel.py [hunk 6]
```diff
@@ -1064,11 +1066,103 @@ def create_group(self):
...
+    def check_resources_limits(self, items_to_upload):
+        if self.need_to_check_service_plan_limits():
+            conn_set = self.get_connection_settings()
+            proxy_settings = self.get_proxy_settings()
+            QUOTA_HANDLE = '/api/component/resource/check_quota'
+            check_quota_url = f'{conn_set.server_url}{QUOTA_HANDLE}'
+            quota_response = requests.post(url=check_quota_url, json=items_to_upload,
+                                           proxies=proxy_settings['proxies'],
+                                           auth=proxy_settings['auth'])
+            if quota_response.status_code != 200:
```
Comment (5Nap) [inline src/tree_panel.py:1118]: 201?

File: src/tree_panel.py [hunk 7]
```diff
@@ -1064,11 +1066,103 @@ def create_group(self):
...
+
+        if plan_response.status_code != 200:
+            qgisLog('Unable to get current plan.', level=CompatQgisMsgLogLevel.Critical)
+            return True
+        if not SERVICE_PLAN_KEY in plan_response.json():
+            qgisLog('Unable to parse plan response.', level=CompatQgisMsgLogLevel.Critical)
+            return True
+
+        service_plan = plan_response.json()[SERVICE_PLAN_KEY]
+        if service_plan == FREE_SERVICE_PLAN:
```
Comment (5Nap) [inline src/tree_panel.py:1106]: What if free plan will be renamed? Fail-safe way is to check all other plans

File: src/tree_panel.py [hunk 8]
```diff
@@ -1064,11 +1066,103 @@ def create_group(self):
...
+        plan_url = f'{conn_set.server_url}{PLAN_HANDLE}'
+
+        SERVICE_PLAN_KEY = 'service_plan'
+        FREE_SERVICE_PLAN = 'lemon'
+        plan_response = requests.get(url=plan_url, proxies=proxy_settings['proxies'], auth=proxy_settings['auth'])
+
+        if plan_response.status_code != 200:
+            qgisLog('Unable to get current plan.', level=CompatQgisMsgLogLevel.Critical)
+            return True
+        if not SERVICE_PLAN_KEY in plan_response.json():
```
Comment (5Nap) [inline src/tree_panel.py:1101]: `if X not in Y`

File: src/tree_panel.py [hunk 9]
```diff
@@ -1064,11 +1066,103 @@ def create_group(self):
...
+        return {
+            'proxies': proxies,
+            'auth': auth
+        }
+
+    def get_connection_settings(self):
+        name_of_conn = NgwPluginSettings.get_selected_ngw_connection_name()
+        return NgwPluginSettings.get_ngw_connection(name_of_conn)
+
+    def need_to_check_service_plan_limits(self):
```
Comment (5Nap) [inline src/tree_panel.py:1088]: Literally, it's `def user_has_a_basic_subscription` and it might be a cached property

File: src/tree_panel.py [hunk 10]
```diff
@@ -1064,11 +1066,103 @@ def create_group(self):
...
+            if isinstance(layer, QgsRasterLayer):
+                items_count['raster_layer'] += 1
+            elif isinstance(layer, QgsVectorLayer):
+                items_count['vector_layer'] += 1
+        if is_project:
+            items_count['webmap'] = 1
+        return items_count
+
+    def display_update_subscription_window(self):
+        url = self.tr('https://nextgis.ru/pricing-base/?source=qgis&amp;utm_source=qgis_plugin&amp;utm_medium=referal&amp;utm_campaign=connect&amp;utm_content=ru')
```
Comment (5Nap) [inline src/tree_panel.py:1147]: I'm not sure if it should be translatable

**Comment by ivanbarsukov** [comment]
Implemented by https://github.com/nextgis/ngw_external_api_python/commit/9a24d36b5fdbf71ecb6e227db1f6b39a65e2a801

**Pull request closed without merging**